src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -51,8 +51,22 @@
 import java.util.stream.Collectors;
 
 /**
- * Scanner that will scan down all forks when we hit parallel blocks before continuing, but generally runs in linear order
- * <p>Think of it as the opposite of {@link DepthFirstScanner}.
+ * Scanner that will scan down all forks when we hit parallel blocks before continuing (as opposed to {@link DepthFirstScanner}), but generally runs in linear order.
+ * <p><strong>Warning</strong> This scanner has various issues when iterating over incomplete builds, or more generally
+ * when multiple heads are passed to {@link #setup}, especially if the build contains nested parallelism.
+ * Consider using {@link DepthFirstScanner} instead in those cases to avoid these issues, employing its iteration order
+ * as needed to sort results (although its global iteration order is inconsistent, its relative iteration order among
+ * siblings at the same nesting level (e.g. parallel branches) is consistent and that is good enough for typical use
+ * cases such as producing a tree).
```

Comment:
Note though that `DepthFirstScanner` has inconsistent behavior with regard to parallel branches: it visits them (and the current heads) first-to-last instead of last-to-first as would be natural given the overall end-to-start iteration order (I think for backwards compatibility with `FlowGraphWalker`). This means that although the relative order among siblings is meaningful, that order is opposite between parallel branches and other kinds of nodes at the same level, so a special case is necessary. This is easy to work around, but perhaps we should add a `DepthFirstScanner(boolean useNaturalOrder)` constructor that would avoid this inconsistency.